keps/sig-multicluster/3335-statefulset-slice/README.md

```diff
@@ -0,0 +1,934 @@
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+The following caveats are applicable to migrating a StatefulSet (scaling down one slice and scaling up another). The following caveats are outside the scope of this KEP, but are applicable to the User Journey of migration motivated by this feature.
+
+**Networking:** Managing services and networking during migration is outside the scope of this proposal. Cross cluster migration can leverage [Multi-Cluster Services](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api) to establish connectivity between pods in different slices. The application operator must set up Multi-Cluster Services in the clusters, and the underlying Stateful application must be configured appropriately. Cross namespace migration can leverage a fallback domain, referring to services from both slices. Similarly this requires an application to be aware of both services.
+
+**Storage:** StatefulSets that use `volumeClaimTemplates`, will create pods that consume per replica PVCs. PVs are cluster scoped resources, but are bound one-to-one with namespace scoped PVCs. If the underlying storage is to be re-used in the new namespace, PVs must be unbound and manipulated appropriately.
+
+**Orchestration:** Consider migrating from namespace `A` to `B`. To preserve StatefulSet at most one semantics, pods should only be migrated when safe to do so. If migrating across namespaces, a pod with ordinal `i` should be scaled down in slice `A` before it is scaled up in slice `B`.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+This KEP proposes a new field `replicaStartOrdinal` with a default value of `0`. StatefulSet will maintain current behavior, if this field is unset.
+
+To mitigate risk, this feature will be rolled out with an alpha feature gate for experimentation. In Beta, new functionality should only take effect if the field `replicaStartOrdinal` is set to a value greater than `0`.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+**StatefulSet Spec Changes**
+
+```
+type StatefulSetSpec struct {
```

Comment:
I slightly prefer:

```
type StatefulSetSpec struct {
        // Ordinals controls how the stateful set creates pod and persistent volume claim names.
        // The default behavior assigns a number starting with zero and incremented by one for each
        // additional replica requested.
        Ordinals struct {
          // The startOrdinal is the number representing the first index
          // that is used to represent replica ordinals
          // If set, replica ordinals will be numbered
          // [replicaStartOrdinal, replicaStartOrdinal - replicas)
          // If unspecified, defaults to 0
          // +optional
          StartOrdinal *int32
}
```

We say ordinals a lot, we can use the description of the struct to better define.

We could also say `ordinalStrategy` but since we're in spec (implied "we're defining how this works") I think it's less necessary.  Also considered `ordinalPolicy` but seems redundant.  We could say `spec.ordinals.start` instead of `startOrdinal` for less repeating.

We could avoid creating the struct now, but then that would require a move and a new API version if we add them, and I think there's a non-zero chance we need more control over ordinals.

Reply:
I like this idea of encapsulating the Start within an Ordinals struct `spec.ordinals.start`. It clearly indicates that there is a distinct dimension to manipulating the ordinals of a StatefulSet.